Add schema design best practices (#1840)
influxdata/docs-v2: content/influxdb/v2.0/write-data/best-practices/resolve-high-cardinality.md

---
title: Resolve high series cardinality
description: >

>   Reduce high series cardinality in InfluxDB. If reads and writes to InfluxDB have started to slow down, you may have high cardinality. Find the source of high cardinality and fix your schema to resolve high cardinality issues.

Author: I agree!

menu:
  influxdb_2_0:
    name: Resolve high cardinality
    weight: 202
    parent: write-best-practices
---

If reads and writes to InfluxDB have started to slow down, high [series cardinality](/influxdb/v2.0/reference/glossary/#series-cardinality) (too many series) may be causing memory issues.

To resolve high series cardinality, complete the following steps (for multiple buckets if applicable):

1. [Review tags](#review-tags).
2. [Adjust your schema](#adjust-your-schema).

## Review tags

Review your tags to ensure each tag **does not contain** unique values for most entries:

- Scan your tags for [common tag issues](#common-tag-issues).
- Use the example Flux query below to [count unique tag values](#count-unique-tag-values).

### Common tag issues

Look for the following common issues, which often cause many unique tag values:

- **Writing log messages to tags**. If a log message includes a unique timestamp, pointer value, or unique string, many unique tag values are created.
- **Writing timestamps to tags**. Typically done by accident in client code.

> - **Tags initially set up with few unique values that grow over time.** For example, a user ID tag may work at a small startup, but may begin to cause issues when the company grows to thousands of users.

Reviewer: ```suggestion
- **Unique values that grow over time set as tags** For example, a user ID tag may work at a small startup, but may begin to cause issues when the company grows to thousands of users.
```

Reworded a bit.

Reviewer: I think I left out the . mark

Author: Thanks, @mprintz! Will reword to: Unique tag values that grow over time.
